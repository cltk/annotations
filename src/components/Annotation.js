// @flow

import React from 'react'

import {
  ANNOTATION_BLUE,
  ANNOTATION_HOVER_BLUE
} from '../constants'

export type Props = {
  annotationHoverEntityKey: string,
  children: string | Object,
  entityKey: string,
  onClick: string => void,
  onMouseOut: string => void,
  onMouseOver: string => void,
  revealedEntityKey: string,
  setCurrentTextState: Function,
}

const Annotation = (
  hoverEntityKey: string,
  onHover: (entityKey: ?string) => void,
  onClick: (entityKey: string) => void
) => {
  return class AnnotationComponent extends React.Component {
    handleClick: Function
    handleMouseOut: Function
    handleMouseOver: Function
    isHoverTarget: void => boolean

    static defaultProps = {
      annotationHoverEntityKey: '',
      onClick: () => {},
      onMouseOut: () => {},
      onMouseOver: () => {},
      revealedEntityKey: '',
    }

    constructor(props: Props) {
      super(props)

      this.handleClick = this.handleClick.bind(this)
      this.handleMouseOut = this.handleMouseOut.bind(this)
      this.handleMouseOver = this.handleMouseOver.bind(this)
      this.isHoverTarget = this.isHoverTarget.bind(this)
    }

    handleClick() {
      onClick(this.props.entityKey)
    }

    handleMouseOut() {
      onHover('')
    }

    handleMouseOver() {
      onHover(this.props.entityKey)
    }

    isHoverTarget() {
      return this.props.entityKey === hoverEntityKey
    }

    render() {
      const style = {
        cursor: 'pointer',
        backgroundColor: this.isHoverTarget() ?
          ANNOTATION_HOVER_BLUE :
          ANNOTATION_BLUE,
      }

      return (
        <span
          onClick={this.handleClick}
          onMouseOut={this.handleMouseOut}
          onMouseOver={this.handleMouseOver}
          role="presentation"
          style={style}
        >
          {this.props.children}
        </span>
      )
    }
  }
}

export default Annotation
