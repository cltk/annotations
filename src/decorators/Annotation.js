// @flow

import type { ContentBlock, ContentState } from 'draft-js'

import {
  ANNOTATION_ENTITY_TYPE,
} from '../constants'
import Annotation from '../components/Annotation'

export const findNoteEntities = (
  contentBlock: ContentBlock,
  callback: Function,
  contentState: ContentState
) => {
  contentBlock.findEntityRanges(
    character => {
      const entityKey = character.getEntity()

      return (
        entityKey !== null &&
        contentState.getEntity(entityKey).getType() === ANNOTATION_ENTITY_TYPE
      )
    },
    callback,
  )
}

export const annotationDecorator = (
  hoverEntityKey: string,
  onHover: (entityKey: string) => void,
  onClick: (entityKey: string) => void,
) => ({
  strategy: findNoteEntities,
  component: Annotation(hoverEntityKey, onHover, onClick)
})


