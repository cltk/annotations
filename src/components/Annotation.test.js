import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import Annotation from './Annotation'

describe('Annotation component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <Annotation onClick={() => {}}>text</Annotation>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('calls `onClick` prop on click', () => {
    const onClick = jest.fn()
    const annotation = shallow(
      <Annotation onClick={onClick}>text</Annotation>
    )
    annotation.find('span').simulate('click')

    expect(onClick).toHaveBeenCalled()
  })

  it('calls `onMouseOut` prop on mouseout', () => {
    const onMouseOut = jest.fn()
    const annotation = shallow(
      <Annotation onMouseOut={onMouseOut}>text</Annotation>
    )
    annotation.find('span').simulate('mouseout')

    expect(onMouseOut).toHaveBeenCalled()
  })

  it('calls `onMouseOver` prop on mouseover', () => {
    const onMouseOver = jest.fn()
    const annotation = shallow(
      <Annotation onMouseOver={onMouseOver}>text</Annotation>
    )
    annotation.find('span').simulate('mouseover')

    expect(onMouseOver).toHaveBeenCalled()
  })
})
