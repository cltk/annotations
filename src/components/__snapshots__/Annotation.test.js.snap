// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Annotation component renders correctly 1`] = `
<span
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  role="presentation"
  style={
    Object {
      "backgroundColor": "#71bed6",
      "cursor": "pointer",
    }
  }
>
  text
</span>
`;
